File: scuttlebutt-client-lib/src/integrationTest/java/org/apache/tuweni/scuttlebutt/lib/FeedStreamTest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more contributor license agreements. See the NOTICE
 * file distributed with this work for additional information regarding copyright ownership. The ASF licenses this file
 * to You under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the
 * License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 */
package org.apache.tuweni.scuttlebutt.lib;

import static org.apache.tuweni.scuttlebutt.lib.Utils.getMasterClient;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;
import static org.junit.jupiter.api.Assumptions.assumeTrue;

import org.apache.tuweni.concurrent.AsyncResult;
import org.apache.tuweni.concurrent.CompletableAsyncResult;
import org.apache.tuweni.crypto.sodium.Sodium;
import org.apache.tuweni.junit.BouncyCastleExtension;
import org.apache.tuweni.junit.VertxExtension;
import org.apache.tuweni.junit.VertxInstance;
import org.apache.tuweni.scuttlebutt.lib.model.FeedMessage;
import org.apache.tuweni.scuttlebutt.lib.model.StreamHandler;
import org.apache.tuweni.scuttlebutt.rpc.RPCAsyncRequest;
import org.apache.tuweni.scuttlebutt.rpc.RPCFunction;
import org.apache.tuweni.scuttlebutt.rpc.RPCResponse;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Map;
import java.util.Optional;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.vertx.core.Vertx;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

@ExtendWith({VertxExtension.class, BouncyCastleExtension.class})
class FeedStreamTest {

  @BeforeAll
  static void checkSodium() {
    assumeTrue(Sodium.isAvailable(), "Sodium native library is not available");
  }

  /**
   * Tests it is possible to make posts and retrieve them again using the FeedService class
   */
  @Test
  void testCreateFeedStream(@VertxInstance Vertx vertx) throws Exception {
    ScuttlebuttClient scuttlebuttClient = getMasterClient(vertx);

    FeedService feedService = scuttlebuttClient.getFeedService();

    AsyncResult<List<FeedMessage>> published = publishTestMessages(feedService);

    // Wait for the messages to be published.
    List<FeedMessage> feedMessages = published.get();

    assertEquals(feedMessages.size(), 10);

    Optional<FeedMessage> lastPosted =
        feedMessages.stream().max(Comparator.comparingLong(msg -> msg.getValue().getSequence()));

    assertTrue(lastPosted.isPresent());

    CompletableAsyncResult<Optional<FeedMessage>> lastMessage = AsyncResult.incomplete();

    feedService.createFeedStream((closer) -> new StreamHandler<>() {

      Optional<FeedMessage> currentMessage = Optional.empty();

      @Override
      public void onMessage(FeedMessage item) {
        currentMessage = Optional.of(item);
      }

      @Override
      public void onStreamEnd() {
        lastMessage.complete(currentMessage);
      }

      @Override
      public void onStreamError(Exception ex) {
        fail(ex.getMessage());
      }
    });

    Optional<FeedMessage> lastStreamedMessage = lastMessage.get();

    assertTrue(lastStreamedMessage.isPresent());

    assertEquals(lastStreamedMessage.get().getValue().getSequence(), lastPosted.get().getValue().getSequence());

    TestScuttlebuttSerializationModel content =
        lastStreamedMessage.get().getValue().getContentAs(new ObjectMapper(), TestScuttlebuttSerializationModel.class);

    assertEquals("serialization-test", content.getType());

    AsyncResult<RPCResponse> result = scuttlebuttClient
        .rawRequestService()
        .makeAsyncRequest(new RPCAsyncRequest(new RPCFunction("whoami"), Collections.emptyList()));
    assertNotNull(result.get());
    @SuppressWarnings("rawtypes")
    Map map = result.get().asJSON(new ObjectMapper(), Map.class);
    assertTrue(map.containsKey("id"));
  }

  private AsyncResult<List<FeedMessage>> publishTestMessages(FeedService feedService) throws JsonProcessingException {

    List<AsyncResult<FeedMessage>> results = new ArrayList<>();
    for (int i = 0; i < 10; i++) {
      AsyncResult<FeedMessage> result = feedService.publish(new TestScuttlebuttSerializationModel("test: " + i));

      results.add(result);
    }

    return AsyncResult.combine(results);
  }


}

File: scuttlebutt-client-lib/src/integrationTest/kotlin/org/apache/tuweni/scuttlebutt/lib/FeedStreamTest.kt
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.tuweni.scuttlebutt.lib

import com.fasterxml.jackson.core.JsonProcessingException
import com.fasterxml.jackson.databind.ObjectMapper
import io.vertx.core.Vertx
import kotlinx.coroutines.runBlocking
import org.apache.tuweni.concurrent.AsyncResult
import org.apache.tuweni.crypto.sodium.Sodium
import org.apache.tuweni.junit.BouncyCastleExtension
import org.apache.tuweni.junit.VertxExtension
import org.apache.tuweni.junit.VertxInstance
import org.apache.tuweni.scuttlebutt.lib.model.FeedMessage
import org.apache.tuweni.scuttlebutt.lib.model.StreamHandler
import org.apache.tuweni.scuttlebutt.rpc.RPCAsyncRequest
import org.apache.tuweni.scuttlebutt.rpc.RPCFunction
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assumptions
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import java.util.Optional

@ExtendWith(VertxExtension::class, BouncyCastleExtension::class)
internal class FeedStreamTest {

  companion object {
    @JvmStatic
    @BeforeAll
    fun checkSodium() {
      Assumptions.assumeTrue(Sodium.isAvailable(), "Sodium native library is not available")
    }
  }

  /**
   * Tests it is possible to make posts and retrieve them again using the FeedService class
   */
  @Test
  @Throws(Exception::class)
  fun testCreateFeedStream(@VertxInstance vertx: Vertx) = runBlocking {
    val scuttlebuttClient = Utils.getMasterClient(vertx)
    val feedService = scuttlebuttClient.feedService
    val feedMessages = publishTestMessages(feedService)

    Assertions.assertEquals(feedMessages.size, 10)
    val lastPosted =
      feedMessages.stream().max(Comparator.comparingLong { (_, _, value): FeedMessage -> value.sequence })
    Assertions.assertTrue(lastPosted.isPresent)
    val lastMessage = AsyncResult.incomplete<Optional<FeedMessage>>()
    feedService.createFeedStream {
      object : StreamHandler<FeedMessage> {
        var currentMessage = Optional.empty<FeedMessage>()
        override fun onMessage(item: FeedMessage) {
          currentMessage = Optional.of(item)
        }

        override fun onStreamEnd() {
          lastMessage.complete(currentMessage)
        }

        override fun onStreamError(ex: Exception?) {
          Assertions.fail<Any>(ex!!.message)
        }
      }
    }
    val lastStreamedMessage = lastMessage.get()!!
    Assertions.assertTrue(lastStreamedMessage.isPresent)
    Assertions.assertEquals(lastStreamedMessage.get().value.sequence, lastPosted.get().value.sequence)
    val content = lastStreamedMessage.get().value.getContentAs(
      ObjectMapper(),
      TestScuttlebuttSerializationModel::class.java
    )
    Assertions.assertEquals("serialization-test", content.type)
    val result = scuttlebuttClient
      .rawRequestService
      .makeAsyncRequest(RPCAsyncRequest(RPCFunction("whoami"), emptyList()))
    assertNotNull(result)
    val map: Map<*, *> = result.asJSON(ObjectMapper(), Map::class.java)
    Assertions.assertTrue(map.containsKey("id"))
  }

  @Throws(JsonProcessingException::class)
  private suspend fun publishTestMessages(feedService: FeedService): List<FeedMessage> {
    val results: MutableList<FeedMessage> = ArrayList()
    for (i in 0..9) {
      val result: FeedMessage = feedService.publish(
        TestScuttlebuttSerializationModel(
          "test: $i"
        )
      )
      results.add(result)
    }
    return results
  }
}
